### PR TITLE
Avoid scan on base table when sync expression is specified on index field

### PR DESCRIPTION
**Pull request by phani-srikar**
#### Description of changes

Adds a check to identify if the current query operation falls outside the delta table TTL window.
Avoids scanning the entire base table but query it when `syncExpression` returns a predicate that maps to a query expression. Relevant [doc](https://docs.amplify.aws/lib/datastore/sync/q/platform/js/#advanced-use-case---query-instead-of-scan).
This improves the performance of the query significantly as called out in the above doc section.

The Delta Sync Table TTL is set to 30 mins by Amplify CLI when conflict detection is enabled by the developer.
Developers however can override this value following the guide [here](https://docs.amplify.aws/cli/graphql/override/#customize-amplify-generated-resources-for-model-directive).
If the delta sync table TTL is overridden, we will use the overridden value in the Sync Query resolver function.

#### Issue #, if available

https://github.com/aws-amplify/amplify-category-api/issues/672

#### Description of how you validated changes
[Successful e2e test run](https://app.circleci.com/pipelines/github/aws-amplify/amplify-category-api/784/workflows/c4d07ec6-0dbb-49eb-a1ad-97db3552189a)

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by alharris-at** [comment]
Is there any way for us to leverage the configured ttl instead of just estimating? I'd prefer us not to ask customers to use custom resolvers if at all possible.

**Comment by phani-srikar** [comment]
> Is there any way for us to leverage the configured ttl instead of just estimating? I'd prefer us not to ask customers to use custom resolvers if at all possible.

Custom overrides are applied centrally in transformer core [after all the stacks and resolvers are generated](https://github.com/aws-amplify/amplify-category-api/blob/7d288436ac4141f122790096f84cd5aff723d718/packages/amplify-graphql-transformer-core/src/transformation/transform.ts#L302) in memory. Tweaking this process to re-run/update the transformers might overwrite the user specified overrides.

**Review by marcvberg — COMMENTED** [review]
Re: Al's comment above - it might make more sense given the context to just pull out those values for TTL into regular CFN parameters, then we could reference them from the resolvers and CFN would put the correct value into the resolvers at deploy time

If a customer configures a higher TTL than default then this might be sub-optimal but no necessarily harmful apart from cost, but if they configure a lower TTL than default without updating this then they'd potentially be working with stale data and could get into more serious issues

At a minimum we would need to document this as a new requirement for customers configuring base table TTL

**Comment by alharris-at** [comment]
> Re: Al's comment above - it might make more sense given the context to just pull out those values for TTL into regular CFN parameters, then we could reference them from the resolvers and CFN would put the correct value into the resolvers at deploy time
> 
> If a customer configures a higher TTL than default then this might be sub-optimal but not necessarily harmful apart from cost, but if they configure a lower TTL than default without updating this then they'd potentially be working with stale data and could get into more serious issues
> 
> At a minimum we would need to document this as a new requirement for customers configuring base table TTL. Possibly even a feature flag considering the impact to existing customers

**Review by marcvberg — APPROVED** [review]
LGTM

**Comment by naveenkoduri** [comment]
> 
> If a customer configures a higher TTL than default then this might be sub-optimal but not necessarily harmful apart from cost, 
Just want to put a perspective as we were bitten by this in April 2022. We originally had default of 30min and ended up with lot of scans despite having the right index for Selective Sync. The scan cost was exponential, and is proportional to the no. of customers using our product. So, at some point, cost would be been a deal breaker and also not consider Amplify Datastore for other products. Our solution was to bump from 30 min to 27 hours for _DeltaSync_ table. Part of our problem was likely due to our configuration of _syncInterval_ of 15 seconds which we commented out falling back to the default of 86400 seconds.

In short, we cannot afford to Scan as our data is huge and devices only want to sync the data that is applicable to them.

**Review by alharris-at — APPROVED** [review]

**Comment by naveenkoduri** [comment]
Hi,  I am working with AWS support, but, thought to leave a comment as a heads up. The changes made with this PR might be missing one use case.  Per the changes made, lastSync is included in the Sync operation if it is not a scan. We do a Selective Sync and our initial sync will use an Index. So, we will not have a value for lastSync for our initial sync to be passed to Sync operation. Hence, it is failing at the client with an error indicating that lastSync must be a number. We do see the transformed template has lastSync=null.
